### PR TITLE
feature: ban the peer when receive misbehave compact-block, add test for compact block process

### PR DESCRIPTION
**Pull request by u2**

### CODE REVIEW

File: sync/src/relayer/compact_block_process.rs [hunk 1]
```diff
@@ -23,6 +23,18 @@ pub struct CompactBlockProcess<'a, CS> {
...
+    UnknownParent,
+    NotBetter,
+    AlreadyInFlight,
+    AlreadyPending,
+    HeaderVerifyFailed,
+    InflightBlocksReachLimit,
+    SendMissingIndexes,
+    NoMissingIndexes,
+}
+
```
Comment (keroro520) [inline sync/src/relayer/compact_block_process.rs]: 1. It is worth a more reasonable name. For example, `SendMissingIndexes` is not a "status", `HeaderVerifyFailed` is an error.
2. Many other places can share this notion. Maybe we should define it as a global type, and use it to help score-system.

File: sync/src/relayer/tests/helper.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
+use crate::{Relayer, SyncSharedState};
```
Comment (keroro520) [inline sync/src/relayer/tests/helper.rs]: YFI, I am planning to move `sycn/src/relayer/tests` into `sync/src/tests/`, after https://github.com/nervosnetwork/ckb/pull/1204.

File: sync/src/relayer/compact_block_process.rs (hunk 1 above)
Comment (u2) [reply sync/src/relayer/compact_block_process.rs]: 1. Naming is hard.
2. Agree. There is too much `Result<(), E>`, it's not good to read, test and debug.

File: sync/src/relayer/tests/helper.rs (hunk 2 above)
Comment (u2) [reply sync/src/relayer/tests/helper.rs]: And can reuse the test utils.

File: sync/src/relayer/compact_block_process.rs (hunk 1 above)
Comment (u2) [reply sync/src/relayer/compact_block_process.rs]: `Status` is the only name which I can think, any suggestion?

Comment (doitian) [reply sync/src/relayer/compact_block_process.rs]: The problem here is that we misuse `Result` here, since we want to ignore some errors. It's better to categorize errors like HTTP, which 4xx are errors caused by clients, and 5xx are caused by the server.

File: sync/src/types.rs [hunk 3]
```diff
@@ -1175,7 +1175,7 @@ mod tests {
     use ckb_core::header::HeaderBuilder;
     use rand::{thread_rng, Rng};
 
-    const SKIPLIST_LENGTH: u64 = 500_000;
```
Comment (doitian) [inline sync/src/types.rs]: This should be in a separate PR.

File: sync/src/relayer/compact_block_process.rs (hunk 1 above)
Comment (u2) [reply sync/src/relayer/compact_block_process.rs]: These are not errors.

**Comment by u2** [comment]
@doitian , @keroro520 updated.

I consider is it necessary to cache the result after verifying the header?
https://github.com/nervosnetwork/ckb/pull/1211/files#diff-98eccb89a5611099d6adca819c9c8b08L100
https://github.com/nervosnetwork/ckb/pull/1211/files#diff-98eccb89a5611099d6adca819c9c8b08L156

**Comment by u2** [comment]
@doitian , @keroro520  updated

File: sync/src/relayer/mod.rs [hunk 4]
```diff
@@ -168,8 +168,10 @@ impl<CS: ChainStore + 'static> Relayer<CS> {
         message: RelayMessage,
     ) {
         if let Err(err) = self.try_process(Arc::clone(&nc), peer, message) {
-            debug_target!(crate::LOG_TARGET_RELAY, "try_process error {}", err);
-            nc.ban_peer(peer, BAD_MESSAGE_BAN_TIME);
+            if let Ok(Error::Misbehavior(e)) = err.downcast() {
+                debug_target!(crate::LOG_TARGET_RELAY, "try_process error {}", e);
+                nc.ban_peer(peer, BAD_MESSAGE_BAN_TIME);
+            }
```
Comment (zhangsoledad) [inline sync/src/relayer/mod.rs]: the original logical will ban malformed message, `Misbehavior` didn't cover this

Comment (u2) [reply sync/src/relayer/mod.rs]: updated

File: sync/src/relayer/error.rs [hunk 5]
```diff
@@ -2,6 +2,24 @@ use failure::Fail;
...
+    #[fail(display = "misbehavior error: {}", _0)]
+    Misbehavior(Misbehavior),
+    #[fail(display = "ignored error: {}", _0)]
+    Ignored(Ignored),
+}
+
+#[derive(Debug, Fail, Eq, PartialEq)]
+pub enum Internal {
+    #[fail(display = "InflightBlocksReachLimit")]
+    InflightBlocksReachLimit,
```
Comment (keroro520) [inline sync/src/relayer/error.rs]: It is better to change the category as below:

* move `InflightBlocksReachLimit` into `Ignore` 
* `MissingIndexesEmpty` is not an `Error`. So I don't think define `MissingIndexesEmpty` is a good idea. If we don't define it, then we can remove the definition of `Internal` (after moving `InflightBlocksReachLimit` into `Ignore` )
* rename `Ignore::BlockStored` -> `Ignore::AlreadyStored`

File: sync/src/relayer/mod.rs [hunk 6]
```diff
@@ -168,8 +169,19 @@ impl<CS: ChainStore + 'static> Relayer<CS> {
...
     ) {
         if let Err(err) = self.try_process(Arc::clone(&nc), peer, message) {
-            debug_target!(crate::LOG_TARGET_RELAY, "try_process error {}", err);
-            nc.ban_peer(peer, BAD_MESSAGE_BAN_TIME);
+            if let Some(&Error::Misbehavior(ref e)) = err.downcast_ref() {
+                debug_target!(crate::LOG_TARGET_RELAY, "try_process error {}", e);
+                nc.ban_peer(peer, BAD_MESSAGE_BAN_TIME);
+                return;
+            }
+            if let Some(&ProtocalError::Malformed) = err.downcast_ref() {
```
Comment (keroro520) [inline sync/src/relayer/mod.rs]: Use `else if` or `match` statement.

File: sync/src/relayer/compact_block_process.rs [hunk 7]
```diff
@@ -197,40 +210,44 @@ impl<'a, CS: ChainStore + 'static> CompactBlockProcess<'a, CS> {
...
-                fbb,
-                &block_hash,
-                &missing_indexes
-                    .into_iter()
-                    .map(|i| i as u32)
-                    .collect::<Vec<_>>(),
+        // The missing_indexes may never be empty
+        if missing_indexes.is_empty() {
+            return Err(Error::Internal(Internal::MissingIndexesEmpty).into());
+        }
```
Comment (keroro520) [inline sync/src/relayer/compact_block_process.rs]: `if missing_indexes.is_empty()` means `Ok(())`, not an error.

File: sync/src/relayer/mod.rs (hunk 6 above)
Comment (u2) [reply sync/src/relayer/mod.rs]: If using `else if`, the Clippy would suggest it can be ignored.
We could no use `match` here, it could not derive the type, the downcast_ref return a generic type.

File: sync/src/relayer/compact_block_process.rs (hunk 7 above)
Comment (u2) [reply sync/src/relayer/compact_block_process.rs]: The `missing_indexes` should never be empty. Otherwise, there would be some logical error in `reconstruct_block`.

File: sync/src/relayer/error.rs (hunk 5 above)
Comment (u2) [reply sync/src/relayer/error.rs]: > It is better to change the category as below:
> 
> * move `InflightBlocksReachLimit` into `Ignore`

I don't want to move it, because in this case, the node could not accept this compact block just because it's busy, the node does not ignore it deliberately.

> * `MissingIndexesEmpty` is not an `Error`. So I don't think define `MissingIndexesEmpty` is a good idea. If we don't define it, then we can remove the definition of `Internal` (after moving `InflightBlocksReachLimit` into `Ignore` )

ok

> * rename `Ignore::BlockStored` -> `Ignore::AlreadyStored`

ok

**Comment by u2** [comment]
@doitian merge?
